qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,378 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" PauliSumOp Class """
+
+import logging
+from numbers import Number
+from typing import Dict, List, Optional, Set, Tuple, Union, cast
+
+import numpy as np
+from scipy.sparse import spmatrix
+
+from qiskit.circuit import Instruction, ParameterExpression
+from qiskit.quantum_info import Pauli, SparsePauliOp
+
+from ... import AquaError
+from ..list_ops.summed_op import SummedOp
+from ..list_ops.tensored_op import TensoredOp
+from ..operator_base import OperatorBase
+from .primitive_op import PrimitiveOp
+
+logger = logging.getLogger(__name__)
+
+
+class PauliSumOp(PrimitiveOp):
+    """Class for Operators backend by Terra's ``SparsePauliOp`` class."""
+
+    def __init__(
+            self,
+            primitive: SparsePauliOp,
+            coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+    ) -> None:
+        """
+        Args:
+            primitive: The SparsePauliOp which defines the behavior of the underlying function.
+            coeff: A coefficient multiplying the primitive.
+
+        Raises:
+            TypeError: invalid parameters.
+        """
+        if not isinstance(primitive, SparsePauliOp):
+            raise TypeError(
+                f"PauliSumOp can only be instantiated with SparsePauliOp, not {type(primitive)}"
+            )
+
+        super().__init__(primitive, coeff=coeff)
+
+    def primitive_strings(self) -> Set[str]:
+        return {"SparsePauliOp"}
+
+    @property
+    def num_qubits(self) -> int:
+        return self.primitive.num_qubits  # type: ignore
+
+    def add(self, other: OperatorBase) -> OperatorBase:
+        if not self.num_qubits == other.num_qubits:
+            raise ValueError(
+                "Sum over operators with different numbers of qubits, {} and {}, is not well "
+                "defined".format(self.num_qubits, other.num_qubits)
+            )
+
+        if isinstance(other, PauliSumOp):
+            return PauliSumOp(
+                self.coeff * self.primitive + other.coeff * other.primitive, coeff=1  # type: ignore
+            )
+
+        from .pauli_op import PauliOp
+
+        if isinstance(other, PauliOp):
+            return PauliSumOp(
+                self.coeff * self.primitive  # type: ignore
+                + other.coeff * SparsePauliOp(other.primitive)
+            )
+
+        return SummedOp([self, other])
+
+    def mul(self, scalar: Union[int, float, complex, ParameterExpression]) -> OperatorBase:
+        if isinstance(scalar, Number) and scalar != 0:
```

Comment:
Rather than needing to import/use Number would it have been better to have checked if not an instance of ParameterExpression?

Reply:
`not isinstance(scalar, ParameterExpression)` can not make sure directly that it can multiplicable with SparsePauliOp.
If you don't like using Number, I want to change `isinstance(scalar, (int, float, complex))`.

Reply:
If its not an instance of ParameterExpression, by virtue of the Union of types there only one of the others is allowed. If something else gets passed in incorrectly by a user, that does not duck type to behave appropriately, things are going to fail somewhere - maybe by checking for Number or int/float/complex it will not fail down that path just down the other one.

Reply:
You are all right. I'm worried about the following. Can the static type checker detect it? (when assuming that mypy knows the class information about SparsePauliOp) This means, is the logic not so complicated?